apps/codelab/src/app/codelabs/angular-thirty-seconds/create-snippet/create-snippet.component.ts

```diff
@@ -1,50 +1,164 @@
-import { Component, OnInit } from '@angular/core';
-import { FormBuilder, Validators } from '@angular/forms';
-import { angularSampleCode } from '../shared/angular-sample';
+import { Component, ElementRef, ViewChild } from '@angular/core';
+import { FormArray, FormBuilder, FormControl, FormGroup, Validators } from '@angular/forms';
+import { MatAutocomplete, MatAutocompleteSelectedEvent, MatChipInputEvent, MatDialog } from '@angular/material';
+import { COMMA, ENTER } from '@angular/cdk/keycodes';
+import { MARKDOWN_PLACEHOLDER, TAGS_LIST, angularSampleCode } from '../shared';
+import { SnippetOverviewComponent } from './snippet-modal/snippet-overview.component';
+import { Observable } from 'rxjs/internal/Observable';
+import { map, startWith } from 'rxjs/operators';
 
 @Component({
   selector: 'codelab-create-snippet',
   templateUrl: './create-snippet.component.html',
-  styleUrls: ['./create-snippet.component.css']
+  styleUrls: ['./create-snippet.component.scss']
 })
-export class CreateSnippetComponent implements OnInit {
+export class CreateSnippetComponent {
 
-  readonly initialCode = angularSampleCode;
-  snippet: any = {};
 
-  readonly snippetForm = this.fb.group({
+  @ViewChild('tagInput') tagInput: ElementRef<HTMLInputElement>;
+  @ViewChild('auto') matAutocomplete: MatAutocomplete;
+
+  TAGS_LIST = TAGS_LIST;
+  tags: Array<string> = ['Tips'];
+  filteredTags: Observable<string[]>;
+
+  snippetForm = this.fb.group({
     title: ['', Validators.required],
-    content: ['', Validators.required],
-    author: [''],
-    level: ['beginner'],
-    tags: [''],
-    bonus: [''],
-    links: [''],
+    level: ['Beginner', Validators.required],
+    tags: ['Tips', Validators.required],
+    content: [MARKDOWN_PLACEHOLDER, Validators.required],
+    bonus: [MARKDOWN_PLACEHOLDER, ''],
+    links: [MARKDOWN_PLACEHOLDER, ''],
     demo: [angularSampleCode]
   });
 
-  constructor(private fb: FormBuilder) {
-    this.snippetForm.valueChanges.subscribe(value => {
-      const raw = this.snippetForm.getRawValue();
-      this.snippet = {
-        ...raw,
-        bonus: raw.bonus.trim(),
-        links: raw.links.trim(),
-        tags: raw.tags.split(',')
-      };
-
-      if (raw.demo['app.component.ts'] !== this.initialCode['app.component.ts']) {
-        this.snippet.componentCode = raw.demo['app.component.ts'];
+  isBonus = false;
+  isLinks = false;
+  isDemo = false;
+
+  separatorKeysCodes: number[] = [ENTER, COMMA];
+
+  constructor(
+    private fb: FormBuilder,
+    public dialog: MatDialog
+  ) {
+    this.filteredTags = this.snippetForm.get('tags').valueChanges.pipe(
```

Comment:
Can this be moved out of component?

Reply:
It was taken exactly from here: https://material.angular.io/components/chips/examples

Please, explain where do you want to see it? :)

Reply:
sorry, I mean constructor